### DISCUSSION TITLE
I have WSI of Breast cancer and I want to run them through custom models and make heatmap

### DISCUSSION THREAD

**Discussion by lomshabhishek** [FAST-Imaging/FAST#181] [Q&A] [answered]
This repo looks like one of the most promising solutions to WSI processing.
Kindly help, Thanks.

**Comment by andreped** [comment]
Hello, @lomshabhishek!

There is a comprehensive tutorial on how to get started using `pyFAST` for processing WSIs in Python [here](https://fast.eriksmistad.no/python-tutorial-wsi.html).
It should contain the necessary details. If you have further questions, feel free to ask :]

If you just care about running your model on new WSIs and visualize the predictions on top of the WSI, I could recommend [FastPathology](https://github.com/AICAN-Research/FAST-Pathology).

**Reply by abhisheklomsh** [reply]
Thanks @andreped for a prompt reply, grateful. 

I have done tiling of a WSI and got my NN to figure out infected tiles, now I want to overlay a heatmap on the original WSI based on the confidence value of NN on those tiles.

I am stuck at the step where I have to stitch together the original WSI from tiles, I may need to send those tile coordinates as annotation back if the stitching is complicated.

**Reply by abhisheklomsh** [reply]
would this work?

import fast

def stitch_tiff_files(input_files, output_file):
  """Stitch together multiple TIFF files into a single file.

  Args:
    input_files: A list of the input TIFF files.
    output_file: The name of the output TIFF file.

  """

  images = []
  for input_file in input_files:
    with open(input_file, "rb") as f:
      image = fast.load_image(f)
    images.append(image)

  stitched_image = fast.stitch_images(images)
  with open(output_file, "wb") as f:
    fast.save_image(stitched_image, f)

if __name__ == "__main__":
  input_files = ["input1.tiff", "input2.tiff", "input3.tiff"]
  output_file = "output.tiff"

  stitch_tiff_files(input_files, output_file)

**Reply by andreped** [reply]
@abhisheklomsh FAST has built in techniques for everything you are trying to do. You do not need to export the result at all to render! 

If you wish to run some classifier model on a WSI and render the predictions on top a WSI as overlay, you can do something like this (I just modified the segmentation example seen in the [tutorial here](https://fast.eriksmistad.no/python-tutorial-wsi.html)):
```
import fast

# Download a patch classifier model from the DataHub
model = fast.DataHub().download('bach-model')

# Download test data (can be removed if you have an image for testing)
fast.downloadTestDataIfNotExists()

importer = fast.WholeSlideImageImporter\
    .create(fast.Config.getTestDataPath() + "/WSI/A05.svs")

tissueSegmentation = fast.TissueSegmentation.create()\
    .connect(importer)

# Generate 256x256 patches at 20X magnification with 10% overlap
generator = fast.PatchGenerator.create(
        256, 256,
        magnification=20,
        overlapPercent=0.1
    ).connect(importer)\
    .connect(1, tissueSegmentation)

classifier = fast.NeuralNetwork.create(
        model.paths[0] + '/pw_classification_bach_mobilenet_v2.onnx',
        scaleFactor=1./255.
    ).connect(generator)

stitcher = fast.PatchStitcher.create()\
    .connect(classifier)

# Display the stitched heatmap result on top of the WSI
renderer = fast.ImagePyramidRenderer.create()\
    .connect(importer)

heatmapRenderer = fast.HeatmapRenderer.create()\
    .connect(stitcher)

fast.SimpleWindow2D.create()\
    .connect(renderer)\
    .connect(heatmapRenderer)\
    .run()
```

Given that you have downloaded the test data (includes the `A05.svs` whole slide image (WSI)), you should be able to run this script - a window should show and predictions should be rendered in real-time on top of the WSI. Try running this script first to see that everything works.

To run a custom model on any WSI, you will need to adjust the script:
- Downloading test data takes a long time. You do not need to download the test data if you already have a WSI for testing. Remove the line if not needed.
- You do not need to download a model from DataHub. Just replace the input model path to the `fast.NeuralNetwork.create()` process object (PO).
- Also remember to update the path to the `fast.WholeSlideImageImporter.create()` PO to a WSI you have locally. The image provided here is part of the `FAST` test data.
- You might also need to change other parameters such as the patch size (here `256 x 256`) and magnification (here `20`) to match what your model expects.

Let me know how it goes :]

**Reply by abhisheklomsh** [reply]
Thanks @andreped 
Can I run this on streamlit or render using matplotlib instead of

```
fast.SimpleWindow2D.create()\
    .connect(renderer)\
    .connect(heatmapRenderer)\
    .run()
```

**Reply by andreped** [reply]
> Can I run this on streamlit or render using matplotlib instead of

Sure. The produced heatmap that is formed by the `fast.PatchStitcher` can be save on disk in the HDF5 format.
The result can then be loaded to a numpy array.

First remove all process objects (POs) after the `PatchStitcher`, i.e., all renderers.
Then add the following PO after the stitcher:
```
exporter = fast.HDF5TensorExporter.create("/path/to/prediction.h5")
exporter.run()
```

You can then load the result into a numpy array, and use it as you wish:
```
import h5py

with h5py.File("/path/to/prediction.h5", "r") as f:
    pred_array = np.asarray(f["tensor"])
```

The `pred_array` variable should be an array of shape height, width, channels, where channels should correspond to number of classes.

Note that this prediction image is a lot smaller than the original WSI, which is due to each image patch corresponding to a single pixel in the prediction image.

Note that you can get a downsampled version of the original WSI and plot it by:
```
import fast
import matplotlib.pyplot as plt
import numpy as np

# Run importer and get data
imagePyramid = fast.WholeSlideImageImporter\
    .create(fast.Config.getTestDataPath() + "/WSI/A05.svs")\
    .runAndGetOutputData()

# Extract specific level as FAST image
image = imagePyramid.getAccess(fast.ACCESS_READ).getLevelAsImage(level=2)

# Convert FAST image to numpy ndarray and plot
image = np.asarray(image)
# This should print: (2664, 3222, 3) uint8 255 0
print(image.shape, image.dtype, np.max(image), np.min(image))

plt.imshow(image)
plt.show()
```

To plot the downsampled WSI together with the heatmap, you can plot them side by side using:
```
import matplotlib.pyplot as plt

fig, ax = plt.subplots(1, 3)
ax[0].imshow(image)
ax[1].imshow(pred[..., 0])
ax[2].imshow(pred[..., 1])
plt.show()
```

Here we assume that your model has two classes. For more, adjust the script accordingly.

Hope it helps! :]

---
EDIT: Regarding StreamLit, we have yet to look into it. 
- FAST does not offer web-based rendering (yet). However, I have successfully demonstrated that `FAST` can be used for processing WSIs in the open web-based software `Digital Slide Archive` **as the backend**. I developed a plug-in for that called [FP-DSA-plugin](https://github.com/andreped/FP-DSA-plugin).
- I have also been looking into using FAST for the backend using `gradio` as the frontend (see [neukit](https://github.com/andreped/neukit) project). However, we have yet to venture into testing StreamLit as an alternative to Gradio.
- Note that you can use `PyQt/PySIde` as an alternative to StreamLit. [here](https://fast.eriksmistad.no/surface_extraction_gui_with_pyside2_8py-example.html) is an example of how to develop your own PySide GUI using FAST.

**Reply by lomshabhishek** [reply]
should I use .connect() as well?

```
exporter = fast.HDF5TensorExporter.create("/path/to/prediction.h5")
exporter.run()
```

**Reply by andreped** [reply]
> should I use .connect() as well?

Yes, try running the script including it and report back if it worked :)

**Reply by lomshabhishek** [reply]
Facing issue with exporter 
kindly help 

 exporter = fast.HDF5TensorExporter.create("prediction.h5")
---> 30 exporter.run()
RuntimeError: Input port 0 on process object HDF5TensorExporter is missing its required connection.

  exporter = fast.HDF5TensorExporter.create("prediction.h5").connect(stitcher)
---> 30 exporter.run()
RuntimeError: Bad cast. Tried to cast from type ImagePyramid to Tensor

**Reply by andreped** [reply]
> RuntimeError: Input port 0 on process object HDF5TensorExporter is missing its required connection.

This happens because you had not connected the stitcher as input to the exporter. After the stitcher in the example above, you should do this:
```
stitcher = fast.PatchStitcher.create()\
    .connect(classifier)

exporter = fast.HDF5TensorExporter.create("/path/to/prediction.h5")\
    .connect(stitcher)
exporter.run()
```

I see that you have done that in the other example. In this case, it sounds like the network is producing a segmentation output, which in turn will result in the stitcher to form an ImagePyramid. If the network is an image classifier, the stitched result should just be a heatmap, which then is compatible with the `fast.HDF5TensorExporter`.

Could you try using this classifier instead, just to sanity check this:
```
classifier = fast.NeuralNetwork.create(
        model.paths[0] + '/pw_classification_bach_mobilenet_v2.onnx',
        scaleFactor=1./255.
    ).connect(generator)
```

**Reply by lomshabhishek** [reply]
I am able to get the segmentation layer by using this piece

```
stitcher = fast.PatchStitcher.create()\
    .connect(segmentation)
finished = fast.RunUntilFinished.create()\
    .connect(stitcher)
exporter = fast.TIFFImagePyramidExporter.create('segmented-nuclei-result.tiff')\
    .connect(finished)\
    .run()
```

However it only saves the segmentation layer.
```
exporter = fast.TIFFImagePyramidExporter.create('segmented-nuclei-result.tiff')\
    .connect(renderer)\
    .connect(segmentationRenderer)\
    .run()
```
RuntimeError: ImagePyramidRenderer has no output port with ID 0

Is not working

**Reply by andreped** [reply]
I adjusted the original example and tested it locally on my macbook. Works fine. Here is a working example:
```
import fast
import h5py
import numpy as np

# enable verbose in FAST
fast.Reporter.setGlobalReportMethod(fast.Reporter.COUT)

# Download test data
fast.downloadTestDataIfNotExists()

# Download a patch classifier model from the DataHub
model = fast.DataHub().download('bach-model')

importer = fast.WholeSlideImageImporter\
    .create(fast.Config.getTestDataPath() + "/WSI/A05.svs")

tissueSegmentation = fast.TissueSegmentation.create()\
    .connect(importer)

# Generate 256x256 patches at 20X magnification with 10% overlap
generator = fast.PatchGenerator.create(
        256, 256,
        magnification=1,
        overlapPercent=0.1
    ).connect(importer)\
    .connect(1, tissueSegmentation)

classifier = fast.NeuralNetwork.create(
        model.paths[0] + '/pw_classification_bach_mobilenet_v2.onnx',
        scaleFactor=1./255.
    ).connect(generator)

stitcher = fast.PatchStitcher.create()\
    .connect(classifier)

finish = fast.RunUntilFinished.create()\
    .connect(stitcher)

exporter = fast.HDF5TensorExporter.create("./heatmap_prediction.h5")\
    .connect(finish)
exporter.run()


# load prediction from disk
with h5py.File("./heatmap_prediction.h5", "r") as f:
    pred_array = np.asarray(f["tensor"])
    spacing = np.asarray(f["spacing"])

print(pred_array.shape)
print(spacing)
```

Note that `fast.RunUntilFinished()` was missing and I reduced the magnification level in the patch generation just to make it faster to run the example.

Try to run it and report back if it worked. Then if it worked, try to use your own model, and then a different WSI.

**Reply by andreped** [reply]
> I am able to get the segmentation layer by using this piece

I was not aware that you had a segmentation model. The output of such a model is not referred to as a heatmap. You can save the result of such a model using the `fast.TIFFImagePyramidExporter` as you did, but you should **not** provide the renderer as input, only the stitcher, like so:
```
exporter = fast.TIFFImagePyramidExporter.create('segmented-nuclei-result.tiff')\
    .connect(stitcher)\
    .run()
```

Note that the output is then an image pyramid, similar to the WSI, and not a heatmap. You will have to use the `fast.TIFFImagePyramidImporter` to load it back to Python.

If you wish to get a low-resolution variant of the segmentation output, you can then do something like:
```
# Load result
imagePyramid = fast.TIFFImagePyramidImporter.create("/path/to/prediction.tiff").runAndGetOutputData()

# Extract specific level as FAST image
image = imagePyramid.getAccess(fast.ACCESS_READ).getLevelAsImage(level=2)

# Convert FAST image to numpy ndarray and plot
image = np.asarray(image)
# This should print: (2664, 3222, 3) uint8 255 0
print(image.shape, image.dtype, np.max(image), np.min(image))
```

You can get different resolution of the prediction image by changing the `level` of the `getLevelAsImage()` method.

**Reply by lomshabhishek** [reply]
![output_r](https://github.com/smistad/FAST/assets/139432055/a8ea1d91-b8ac-4ebb-85c3-4312bfd83e7c)
I am getting this output

**Reply by andreped** [reply]
You likely are missing `fast.RunUntilFinished`:
```
stitcher = fast.PatchStitcher.create()\
    .connect(network)

finish = fast.RunUntilFinished.create()\
    .connect(stitcher)

exporter = fast.TIFFImagePyramidExporter.create('segmented-nuclei-result.tiff')\
    .connect(finish)\
    .run()
```

Note that the `finish` output is connected to the exporter.

Other than that, there are numerous reasons to why the result might be blank (or appear blank). Here are some suggestions you could try:
* It might just be that you need to adjust the dtype or intensity range to see the image. Which values were printed after loading?
* The model made predictions when rendering it as overlay, right? If not, the preprocessing is likely wrong. Did you remember to change the `magnification` to 20 or any appropriate value in the `fast.PatchGenerator` PO?
* Are you doing any zero-mean standardization or intensity normalization in preprocessing. The above example assumes that the image patches were normalized to range [0, 1]. This you can see for the `NeuralNetwork` PO (`scaleFactor=1./255.`). If you do not do any, set `scaleFactor=1.0`
* Since, you are using a segmentation model, you likely want to replace the `fast.NeuralNetwork` PO with the `fast.SegmentationNetwork` PO.

**Reply by lomshabhishek** [reply]
I have got the segmentation layer on purple background, but I wanted to see the segmentation layer on original image.
While rendering it overlays on original image but when exported it uses purple background.

**Reply by andreped** [reply]
Yes, but the image you showed is all purple, indicating that there does not appear to be any detections. Given that you have a breast cancer model, the pipeline says that no cancer was detected. So there is likely something wrong. See my suggestions above for how to proceed.

When that is working then we can implement something to add the tumor predictions as overlay on top of the original WSI.

**Reply by lomshabhishek** [reply]
![Screenshot 2023-07-24 170816](https://github.com/smistad/FAST/assets/139432055/1ae7cd46-eac4-4920-b8d9-f216335dffeb)

I have got the predictions, need these to masked over original image and save the image offline.

**Reply by andreped** [reply, marked as answer]
Great! In that case, the final step is using `matplotlib`. Hence, you have managed to do all relevant steps with `pyFAST`.

See my comment above for how to plot the WSI and prediction image side-by-side: https://github.com/smistad/FAST/discussions/181#discussioncomment-6511505

To plot the generated segmentation as overlay on top of the low-resolution WSI, see [this thread](https://stackoverflow.com/questions/57576686/how-to-overlay-segmented-image-on-top-of-main-image-in-python).

Note that to plot segmentation on top of the low-res WSI, they likely need to be of equal image size. This can be achieved using `OpenCV`'s [resize function](https://www.tutorialkart.com/opencv/python/opencv-python-resize-image/#gsc.tab=0).

To save a figure in matplotlib, see [savefig](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.savefig.html).

As the final steps are not FAST-related, I will let you try that on your own. Best of luck! :]

As you are now able to generate and view the heatmap using pyFAST, which was the original question, please accept my answer and resolve this Discussion thread.

If you have any future FAST-related questions or issues, please make a new Discussion.

**Reply by lomshabhishek** [reply]
Thanks a lot @andreped, you have been a great help :)

**Reply by andreped** [reply]
> Thanks a lot @andreped, you have been a great help :)

Thanks, @lomshabhishek! :] Always fun to help!

If you have any other FAST-related issues, be sure to make a new Discussion.
I have dedicated some time to assist developers in using FAST for pathology applications, so don't be a stranger ;)

**Comment by lomshabhishek** [comment]
import fast

### Download a nuclei segmentation model from the DataHub
model = fast.DataHub().download('nuclei-segmentation-model')

importer = fast.WholeSlideImageImporter.create(
     image_path)

tissueSegmentation = fast.TissueSegmentation.create()\
    .connect(importer)

generator = fast.PatchGenerator.create(
        256, 256,
        magnification=20,
        overlapPercent=0
    ).connect(importer)\
    .connect(1, tissueSegmentation)

segmentation = fast.SegmentationNetwork.create(
        model.paths[0] + '/high_res_nuclei_unet.onnx',
        scaleFactor=1./255.
    ).connect(generator)
stitcher = fast.PatchStitcher.create()\
    .connect(segmentation)
finished = fast.RunUntilFinished.create()\
    .connect(stitcher)



exporter = fast.TIFFImagePyramidExporter.create('segmented-nuclei-result_2.tiff')\
    .connect(importer)\
    .connect(finished)\
    .run()


import fast
import matplotlib.pyplot as plt
import numpy as np

### Run importer and get data

imagePyramid = fast.TIFFImagePyramidImporter\
    .create('segmented-nuclei-result_2.tiff')\
    .runAndGetOutputData()

### Extract specific level as FAST image
image = imagePyramid.getAccess(fast.ACCESS_READ).getLevelAsImage(level=2)

### Convert FAST image to numpy ndarray and plot
image = np.asarray(image)
### This should print: (2664, 3222, 3) uint8 255 0
print(image.shape, image.dtype, np.max(image), np.min(image))

plt.imshow(image)
plt.show()

**Reply by Peregalli** [reply]
Hi! 
I have a similar problem. I am using the colon epithelium segmentation HE model and running with my WSI. I would like to get the predictions as a probability (or a heatmap) without output thresholding and save it as a tiff. I have been able to save a  segmentation output as a .tiff file and get a segmentation image. But when I import .tiff file is already binary. I have seen that there is an attribute  `heatmapOutput` in `SegmentationNetwork` class but when I set that parameter as True, I can not save the segmentation correctly. Another thing is that I am using `getPatchAsImage` insted of `getLevelAsImage`. I also run this code but the output shape is (14541, 8562, 1).

**Reply by andreped** [reply]
Hello, @Peregalli !

It would be great if you made a separate Q/A Discussion, as this topic is slightly different and this current Discussion has already been resolved.
